internal/blockchain/utxoviewpoint.go

```diff
@@ -755,8 +877,16 @@ func (view *UtxoViewpoint) fetchInputUtxos(block *dcrutil.Block,
 	// the block.  This applies to both transactions earlier in the stake tree
 	// as well as those in the regular tree.
 	for _, stx := range block.STransactions() {
-		isVote := stake.IsSSGen(stx.MsgTx())
-		for txInIdx, txIn := range stx.MsgTx().TxIn {
+		// Ignore treasurybases and treasury spends since they have no inputs.
+		msgTx := stx.MsgTx()
+		isTreasuryBase := isTreasuryEnabled && stake.IsTreasuryBase(msgTx)
+		isTreasurySpend := isTreasuryEnabled && stake.IsTSpend(msgTx)
+		if isTreasuryBase || isTreasurySpend {
```

Comment:
I'm a bit torn on the first one to be honest.  I did consider it for the first case, because the stake identification functions are indeed not very efficient currently, but I was a bit hesitant because solely relying on the index would introduce an implicit ordering assumption that code prior to this has already validated that the first transaction is actually the required `treasurybase` and the rest are not.  That is a true assumption in this particular case as well as the other one mentioned in on of your other comments.

In general, I prefer to avoid as many hidden implicit ordering assumptions like that as possible in consensus except in cases where it really has a huge impact on performance, but I do believe this is a good candidate given it applies to every single stake transaction and the stake identification functions currently do quite a bit of work.

That said, I would say it should be in addition to the check as opposed to in lieu of it (so that it only checks for the first transaction) to act as an additional bulwark.  In order words, this seems reasonable to cover the bases:
```Go
isTreasuryBase := isTreasuryEnabled && txIdx == 0 && stake.IsTreasuryBase(msgTx)
```

For the second case, I did consider that, as it's done in some other parts of the code (e.g. `mempool`) but I decided against it here since unlike the other cases, it necessarily could only realistically possibly avoid the check on a single transaction here versus adding another conditional branch for every other transaction.  We'd need to benchmark it to really see if it's worth it, but my gut tells me it would probably be a wash and ultimately with more efficient stake identification would likely be worse.

Sound reasonable?

Reply:
Yeah.  Although for the second case, I think it would be something minor like the following short-circuit:

```go
isTreasurySpend := isTreasuryEnabled && !isTreasuryBase && stake.IsTSpend(msgTx)
```

But it only saves a few microseconds for the case where it is the treasurybase.

Reply:
I went ahead and modified it to the following which covers both cases, I believe makes the assumption more clear, and avoid the additional conditional branch that would be introduced by checking it twice:

```Go
		shouldBeTreasuryBase := isTreasuryEnabled && txIdx == 0
		if shouldBeTreasuryBase && stake.IsTreasuryBase(msgTx) {
			continue
		}
		isTreasurySpend := isTreasuryEnabled && stake.IsTSpend(msgTx)
		if isTreasurySpend {
			continue
		}
```